packages/mui-base/src/unstable_useNumberInput/useNumberInput.ts

```diff
@@ -0,0 +1,385 @@
+  UseNumberInputInputSlotProps,
+  UseNumberInputIncrementButtonSlotProps,
+  UseNumberInputDecrementButtonSlotProps,
+  UseNumberInputReturnValue,
+} from './useNumberInput.types';
+import { clamp, isNumber } from './utils';
+import extractEventHandlers from '../utils/extractEventHandlers';
+
+type StepDirection = 'up' | 'down';
+
+const STEP_KEYS = ['ArrowUp', 'ArrowDown', 'PageUp', 'PageDown'];
+
+const SUPPORTED_KEYS = [...STEP_KEYS, 'Home', 'End'];
+
+function parseInput(v: string): string {
+  return v ? String(v.trim()) : String(v);
+}
+
+/**
+ *
+ * Demos:
+ *
+ * - [Number Input](https://mui.com/base/react-number-input/#hook)
+ *
+ * API:
+ *
+ * - [useNumberInput API](https://mui.com/base/react-number-input/hooks-api/#use-number-input)
+ */
+export default function useNumberInput(
+  parameters: UseNumberInputParameters,
+): UseNumberInputReturnValue {
+  const {
+    min,
+    max,
+    step,
+    shiftMultiplier = 10,
+    defaultValue: defaultValueProp,
+    disabled: disabledProp = false,
+    error: errorProp = false,
+    onBlur,
+    onInputChange,
+    onFocus,
+    onChange,
+    required: requiredProp = false,
+    readOnly: readOnlyProp = false,
+    value: valueProp,
+    inputRef: inputRefProp,
+    inputId: inputIdProp,
+  } = parameters;
+
+  // TODO: make it work with FormControl
+  const formControlContext: FormControlState | undefined = useFormControlContext();
+
+  const { current: isControlled } = React.useRef(valueProp != null);
+
+  const handleInputRefWarning = React.useCallback((instance: HTMLElement) => {
+    if (process.env.NODE_ENV !== 'production') {
+      if (instance && instance.nodeName !== 'INPUT' && !instance.focus) {
+        console.error(
+          [
+            'MUI: You have provided a `slots.input` to the input component',
+            'that does not correctly handle the `ref` prop.',
+            'Make sure the `ref` prop is called with a HTMLInputElement.',
+          ].join('\n'),
+        );
+      }
+    }
+  }, []);
+
+  const inputRef = React.useRef<HTMLInputElement>(null);
+  const handleInputRef = useForkRef(inputRef, inputRefProp, handleInputRefWarning);
+
+  const inputId = useId(inputIdProp);
+
+  const [focused, setFocused] = React.useState(false);
+
+  // the "final" value
+  const [value, setValue] = React.useState(valueProp ?? defaultValueProp);
+  // the (potentially) dirty or invalid input value
+  const [dirtyValue, setDirtyValue] = React.useState<string | undefined>(
+    value ? String(value) : undefined,
+  );
+
+  React.useEffect(() => {
+    if (!formControlContext && disabledProp && focused) {
+      setFocused(false);
+
+      // @ts-ignore
+      onBlur?.();
+    }
+  }, [formControlContext, disabledProp, focused, onBlur]);
+
+  const handleFocus =
+    (otherHandlers: Record<string, React.EventHandler<any> | undefined>) =>
+    (event: React.FocusEvent<HTMLInputElement>) => {
+      // Fix a bug with IE11 where the focus/blur events are triggered
```

Comment:
Do we care about IE11 anymore? cc @michaldudak

Reply:
Nope.